data-prepper-plugins/saas-source-plugins/jira-source/src/main/java/org/opensearch/dataprepper/plugins/source/jira/JiraIterator.java

```diff
@@ -40,12 +45,41 @@ public JiraIterator(final JiraService service,
     public boolean hasNext() {
         if (firstTime) {
             log.trace("Crawling has been started");
-            itemInfoQueue = service.getJiraEntities(sourceConfig, lastPollTime);
+            startCrawlerThreads();
             firstTime = false;
         }
+        int timeout = HAS_NEXT_TIMEOUT;
+        while (isCrawlerRunning() && itemInfoQueue.isEmpty() && timeout > 0) {
+            try {
+                log.trace("Waiting for crawler queue to be filled for next {} seconds", timeout);
+                Thread.sleep(crawlerQWaitTimeMillis);
+                timeout--;
+            } catch (InterruptedException e) {
+                log.error("An exception has occurred while checking for the next document in crawling queue");
+                Thread.currentThread().interrupt();
+            }
+        }
         return !this.itemInfoQueue.isEmpty();
     }
 
+    private boolean isCrawlerRunning() {
+        boolean isRunning = false;
+        if (!futureList.isEmpty()) {
+            for (Future<Boolean> future : futureList) {
+                if (!future.isDone()) {
+                    isRunning = true;
+                    break;
+                }
+            }
+        }
+        return isRunning;
+    }
+
+    private void startCrawlerThreads() {
+        futureList.add(crawlerTaskExecutor.submit(() ->
+            service.getJiraEntities(sourceConfig, lastPollTime, itemInfoQueue), false));
+    }
+
```

Comment:
These changes look good but I do not see corresponding changes to the JavaIteratorTest.java file. When using futures testing is very important as we realized with Lambda code. Please make sure all these changes have corresponding tests